### DISCUSSION TITLE
Support querying of AAS via SPARQL

### DISCUSSION THREAD

**Discussion by kenwenzel** [linkedfactory/linkedfactory-pod#45] [General]
We should support the querying of AAS via the HTTP API defined in
https://industrialdigitaltwin.org/wp-content/uploads/2023/04/IDTA-01002-3-0_SpecificationAssetAdministrationShell_Part2_API.pdf

The idea is to use a special federated service for IRIs prefixed by "aas-api:":

```
prefix aas: <https://admin-shell.io/aas/3/0/>
select ?sm ?element ?p ?o {
  service <aas-api:https://v3.admin-shell-io.com> {
    { select ?shell { <aas-api:endpoint> <aas-api:shells> ?shell } limit 2 }

    ?shell aas:submodels [ <aas-api:resolved> ?sm ] . ?sm aas:semanticId ?semId . ?semId aas:keys ?key .
    ?key aas:value "https://admin-shell.io/zvei/nameplate/1/0/Nameplate" .
    ?sm (!<:>)+ ?element .

    { ?element a aas:Property } union { ?element a aas:MultiLanguageProperty }
    ?element ?p ?o .
  }
}
```

**Comment by kenwenzel** [comment]
Conversion of key elements to IRIs:
```
prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
prefix aas: <https://admin-shell.io/aas/3/0/>
base <aas://>
select * {
  { select distinct ?keys ?keyType ?keyValue {
    service <aas-api:https://v3.admin-shell-io.com> {
      { select distinct ?sm ?keys {
          { select ?shell { <aas-api:endpoint> <aas-api:shells> ?shell } limit 2 }
          ?shell aas:submodels [ <aas-api:resolved> ?sm ] . ?sm (!<:>)* ?element . ?element aas:keys ?keys .
        }}
      ?keys aas:type ?keyType ; aas:value ?keyValue ; aas:index ?index .
    }
  } order by ?keys ?index}
  
  bind(concat("(", ?keyType, ")", ?keyValue) as ?typedKey) .

  bind(regex(?keyValue, """^[A-Za-z][\\d+-.A-Za-z]*:[^\u0000-\u0020"<>\\^`{|}]*$""") as ?isIri)
  bind((?keyType in ("GlobalReference", "Submodel")) as ?convertToIri) 
  
  bind(
    if (?convertToIri, 
      if (?isIri, ?keyValue, iri(encode_for_uri(?keyValue))),
      concat(if (?lp = rdf:_2, "#", "."), encode_for_uri(?keyValue))
  ) as ?iriPart)
}
```

**Comment by VladimirAlexiev** [comment]
@kenwenzel I commiserate with you! :-)

Everything in these queries cries "Help!".
- Parsing and constructing URLs on the fly (gathering parts with Group By),
- checking whether something is a URL by its pattern,
- checking some `value` against a string instead of having a normal type `a <https://admin-shell.io/zvei/nameplate/1/0/Nameplate>`
- And the killer trick: `!<:>`. Is that some negated wildcard property path? I thought I know SPARQL...

I'm curious how ChatGPT would explain these queries :-) ok, not bad in general...
> This expression retrieves a key (?key) linked to the semantic ID, filtering out specific relationships using the !<:> operator which excludes specific predicates.

**Comment by kenwenzel** [comment]
@VladimirAlexiev The second query can be used to construct IRIs for lists with `aas:keys`.
It was based on our first idea to use existing IRIs as they are and append local identifiers (short IDs) to a base IRI.

The problem with this approach is that some external data sets (like EClass) use a very specific mapping to convert IRDIs to IRIs.
That would require to consider all those specifics within a mapping algorithm for converting keys to IRIs.
